Hypthesis: The API will be more likely to give comments containing mysogynistic comments as toxic compared to feminist comments. 

Disclaimer: I used chatGPT and public data to create the statements. 

In [1]:
from googleapiclient import discovery
import json

API_KEY = 'AIzaSyCFPx5lwn8mMbYPxKgNUVN7iAzTUgvShd4'

client = discovery.build(
  "commentanalyzer",
  "v1alpha1",
  developerKey=API_KEY,
  discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
  static_discovery=False,
)

analyze_request = {
  'comment': { 'text': 'Women are too emotional to be leaders.' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 38, 'score': {'value': 0.3047732, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.3047732, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


In [2]:
analyze_request = {
  'comment': { 'text': 'Girls should stick to traditional roles like cooking and cleaning.' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 66, 'score': {'value': 0.3666224, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.3666224, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


In [3]:
analyze_request = {
  'comment': { 'text': 'Women belong in the kitchen, not in the boardroom.' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 50, 'score': {'value': 0.57271194, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.57271194, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


In [4]:
analyze_request = {
  'comment': { 'text': "She's probably on her period, that's why she's so cranky" },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 56, 'score': {'value': 0.43230394, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.43230394, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


In [5]:
analyze_request = {
  'comment': { 'text': "She's only successful because she fucked her boss." },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 50, 'score': {'value': 0.9029226, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.9029226, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


With the mysogynistic comments the API gave them a score of 0.3047732, 0.3666224, 0.57271194, 0.43230394, 0.37751234.
The average score of these statements is 0.410984764, which is considered non toxic by the API which means the null hypothesis is true at this point. 

In [6]:
analyze_request = {
  'comment': { 'text': 'Equal pay for equal work is a basic human right.' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 48, 'score': {'value': 0.012063419, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.012063419, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


In [7]:
analyze_request = {
  'comment': { 'text': 'Girls can do anything boys can do, and sometimes even better.' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 61, 'score': {'value': 0.23415339, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.23415339, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


In [8]:
analyze_request = {
  'comment': { 'text': 'Feminism is about achieving gender equality, not female superiority.' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 68, 'score': {'value': 0.19870108, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.19870108, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


In [9]:
analyze_request = {
  'comment': { 'text': "Supporting women's rights benefits everyone in society, not just women." },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 71, 'score': {'value': 0.035691574, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.035691574, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


In [10]:
analyze_request = {
  'comment': { 'text': 'Every woman has the right to make decisions about her own body.' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 63, 'score': {'value': 0.034277402, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.034277402, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


With the feminist comments the API gave them the score of 0.034277402, 0.035691574, 0.19870108, 0.23415339, 0.012063419.

The average of these values is .102977373. With this average we can see that feminist comments are considered less toxic compared to mysogynistic comments, however I wanted to test if profanity had anything to do with the rating so I made two more of each but this time using profanity. 

In [11]:
analyze_request = {
  'comment': { 'text': 'I would never let my daughther dress like a whore.' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 50, 'score': {'value': 0.7856813, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.7856813, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


In [16]:
analyze_request = {
  'comment': { 'text': "I have never met a woman who wasn't a cunt" },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 42, 'score': {'value': 0.78855824, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.78855824, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


In [ ]:
analyze_request = {
  'comment': { 'text': 'Kill all men' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 12, 'score': {'value': 0.85333383, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.85333383, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en', 'sv']}


In [19]:
analyze_request = {
  'comment': { 'text': 'Men are so much fucking worse than women' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(response)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 40, 'score': {'value': 0.911035, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.911035, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


With the addition of profane and extreme language the API gave the scores of 0.911035, 0.85333383, 0.78855824, 0.7856813
These statements have an average of 0.8346520925.


What biases do you think might exist in the model based on intuitions or public documentation about how the model was created?
Bias in models can arise from various sources. My hypothesis for the bias of the model is that it weighs words that have a sociately determined negative connotation as toxic and words that are neutral or positive as non toxic. This means that if you say something extremely toxic but use neutral or happy language the model will not see it as a toxic statement because of the biases that were instilled in the model by the developers. 

What were your results?
My results showed that both mysogynistic and feminist comments are considered non-toxic(below .5) as long as you do not use extreme language in the comment. Read my analysis above for more information about the averages and tendecies. 

What theories do you have about why your results are what they are?

Based on my findings I believe the model is more biased towards curse words or extreme words like kill to be toxic. When I made my comments I tried to find ones that were neutral and didn't use negative words and with these comments the model found both mysogynistic comments and feminist comments to be below .5 or non toxic. However, with the addition of profanity the score shot up and became extremely toxic. These results make me think that the model has a toxicity score for every word and then bases the results from that. 